### DISCUSSION TITLE
Turning on `isRecording` outside of an individual `XCTestCase`

### DISCUSSION THREAD

**Discussion by simonbromberg** [pointfreeco/swift-snapshot-testing#457] [Q&A] [answered]
Is there any way currently to set `isRecording` outside of an individual `XCTestCase`  subclass? For example, say I make a change that I know is going to affect a whole bunch of snapshots, currently I would need to find all those test files and set `isRecording` to `true` on all of them (or use the `recording` parameter).

I would like to just be able to turn something on in say an environment variable or in the `AppDelegate` or someplace else. 

I would also prefer to do this without making a subclass of `XCTestCase` or adding a protocol, or making my own version of `assertSnapshot` all of which would require people to remember to use them.

This is especially useful for whenever there's an iOS update and all your snapshots need to be updated.

**Comment by carsten-wenderdel** [comment, marked as answer]
`isRecording` is already a global variable, so you have to set it only once.

To make sure that this happens before any test runs, you could create a `principal class` as described here: https://developer.apple.com/documentation/xctest/xctestobservationcenter

It could look like this:

```swift
@objc(PrincipalClass)
class PrincipalClass: NSObject {
    override init() {
        isRecording = true
    }
}
```

**Reply by simonbromberg** [reply]
Took me a while but yea this does actually work pretty well.

Can also use swizzling to replace the XCTestCase `invokeTest` method to use the newer method of controlling recording in this library.

Though I haven't come up with a solution that also covers Swift Testing in the same way.

**Reply by simonbromberg** [reply]
That being said it seems you can achieve this more simply by setting `SNAPSHOT_TESTING_RECORD` in your environment variables now

**Comment by doozMen** [comment]
When you run test using `swift test` and avoid using objc then I have this small class that I use so you can set an environment variable to actually update the tests.

Do you think this would be something to add in a PR?

```swift
import SnapshotTesting
import XCTest
import InlineSnapshotTesting

/// Convenience that looks for an environment variable  SNAPSHOT_TESTING_UPDATE_TESTS that can be set to true in order to
/// have the tests update from commandline
open class XCTestCaseSnapshot: XCTestCase {
  open override class func setUp() {
    super.setUp()
    SnapshotTesting.diffTool = "ksdiff"
    InlineSnapshotTesting.diffTool = SnapshotTesting.diffTool
    let isRecording = Bool(ProcessInfo.processInfo.environment["SNAPSHOT_TESTING_UPDATE_TESTS"] ?? "false") ?? false
    SnapshotTesting.isRecording = isRecording
    InlineSnapshotTesting.isRecording = isRecording
  }
}

```
Running the tests from cli would then be 

`export SNAPSHOT_TESTING_UPDATE_TESTS=true && swift test`

Let me know what you think?

**Reply by doozMen** [reply]
You can also pass this environment via xcode using the scheme....

**Reply by simonbromberg** [reply]
While this is certainly a valid approach, it relies on subclassing XCTestCase and ensuring all developers remember to use that subclass in all XCTestCase subclasses used in your project.
